float32 s
float32 l